packages/smithy-aws-core/src/smithy_aws_core/user_agent.py

```diff
@@ -0,0 +1,269 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  SPDX-License-Identifier: Apache-2.0
+#  pyright: reportMissingTypeStubs=false,reportUnknownMemberType=false
+
+import os
+import platform
+from string import ascii_letters, digits
+from typing import NamedTuple, Optional, Self, Union, List
+
+from smithy_http.aio.crt import HAS_CRT
+
+_USERAGENT_ALLOWED_CHARACTERS = ascii_letters + digits + "!$%&'*+-.^_`|~"
+_USERAGENT_ALLOWED_OS_NAMES = (
+    "windows",
+    "linux",
+    "macos",
+    "android",
+    "ios",
+    "watchos",
+    "tvos",
+    "other",
+)
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {"darwin": "macos"}
+_USERAGENT_SDK_NAME = "aws-sdk-python"
+
+
+class UserAgentComponent(NamedTuple):
```

Comment:
This seems more appropriate to use a dataclass with. botocore probably couldn't due to versioning constraints. Doing something like:

```python
@dataclass(frozen=True, slots=True)
class UserAgentComponent:
    prefix: str
    name: str
    value: str | None = None
```

Would keep the usage patterns the same and shouldn't have a notable impact on speed. The one thing you don't get is unpacking or integer indexing, but I don't see that being done here anyway.